ROADMAP.md

```diff
@@ -0,0 +1,75 @@
+# libp2p daemon roadmap
+
+Revision: r1; 2018-09-25
+
+Authors (a-z):
+  - @bigs
+  - @raulk
+  - @stebalien
+  - @vyzo
+
+## Introduction
+
+The *libp2p daemon* provides a standalone deployment of a libp2p host, running
+in its own OS process and installing a set of virtual endpoints for co-local
+applications to interact with it. The daemon is written in Go and can therefore
+leverage the vast ecosystem of go-libp2p modules.
+
+By running a *single instance* of the daemon per machine, co-local applications
+can leverage the libp2p stack to communicate with peers, interact with the DHT,
+multiplex protocols, etc. no matter the language they are developed in, nor
+whether a native libp2p implementation exists in that language.
+
+When establishing connections, the daemon handles transport selection, security
+negotiation, and protocol and stream multiplexing. Streams are mapped 1:1 to
+local endpoints (e.g. unix sockets, shared-memory). Writes and reads to/from
+those endpoints are converted to writes and reads to/from the stream, allowing
+any application to interact with a libp2p network through simple, local IO.
+
+The daemon exposes a control endpoint for management, supporting basic
+operations such as peer connection/disconnection, stream opening/closing, etc.
+as well as operations on higher-level subsystems like the DHT, Pubsub, and
+circuits.
+
+Even though user applications can interact with the daemon in an ad-hoc manner,
+we encourage the development of bindings in different languages. We envision
+these as small, lightweight libraries that encapsulate the control calls and the
+local transport IO, into clean, idiomatic APIs that enable the application not
+only to demand actions from the daemon, but also to plug in protocol handlers
+through the native constructs of that language.
+
+A [Gerbil binding](https://github.com/vyzo/gerbil-libp2p/) has been developed,
+and a Go binding is in the works.
+
+## Short-term roadmap
+
+These are the short-term priorities for us. If you feel something is missing,
+please open a [Github issue](https://github.com/libp2p/go-libp2p-daemon/issues).
+
+- Protobuf control API exposed over a Unix domain socket.
+- Connection lifecycle: connecting and disconnecting with peers.
+- Stream lifecycle: opening and closing streams.
+- Stream <> unix socket 1:1 mapping.
+- Daemon identity: auto-generated, and persisted.
+- DHT operations.
+- Pubsub interactions.
```

Comment:
dht and pubsub are coming next, hopefully this week.

Reply:
Sweet, adding constructions signs 🚧 to these!

Reply:
Just one thing came to my mind: We used RPC for `Validator` in pubsub to send content to the Python side to get the validation result. If the pubsub in libp2p daemon wants to support content validation, should it provide a registration function for users to register a callback Unix domain socket?

Reply:
Good point about the validator, i was also thinking about it.
I guess we could open a unix socket, register a validator with a control op, and keep it open to upcall for validation.

Reply:
Good point! Do we want to be able to set DHT validators too? Or are we always going with the default Namespaced+PublicKeyValidator?

https://github.com/libp2p/go-libp2p-kad-dht/blob/a303c4537172e4195887f602f36f6bd5e3cde0b4/dht.go#L56

Reply:
It's not as critical I think, but we can consider it in the longer term.